### DISCUSSION TITLE
Tooltips for modulators

### DISCUSSION THREAD

**Discussion by puffer3** [RomanKubiak/ctrlr#440] [Q&A] [answered]
Hello all! I am happy to ask my first questions to the ctrlr community.  

I am building a panel for a friend who wants me to use tooltips.

I see tooltip format settings on the main panel but I am not sure how to implement them.  Is it simple to do?

Is there a hub for documentation about ctrlr that I havent been able to find?  I mostly search the dead forum and here for information, but a list of modulator attributes or something would really be handy.

Also, I am still using the build I downloaded from the Ctrlr.org. I just found that the github has a newer version that I will have to try to build with xcode when I am back at my desktop (working on a mojave macbook rn).  Image resources will not load on Mac in my version (5.3.198), so I am loading them in 5.3.163 and then reopening in the newer version which is a pain.  Has this been resolved?

**Comment by damiensellier** [comment]
Hi Henry, welcome to the ctrlr community.
The tooltip text settings in the global panel is about the "bubble" displaying the sliders/potentiometers' values when dragging them.

![Bubble](https://user-images.githubusercontent.com/55164385/156851463-ab517463-a75a-4a0e-8538-9d931dbf5a8d.png)

There's no custom [tooltips class from JUCE](https://docs.juce.com/master/classTooltipWindow.html) implemented in Ctrlr yet. It would require having a LUA script triggered on mouseover.
![mouseover script trigger](https://user-images.githubusercontent.com/55164385/156851521-265bd00b-32e5-42b7-810d-66fe03e56e29.png)


Absolutely, the 5.3.198 has its source image dropdown selector broken.
I would recommand to download the latest working version 5.6.23 here : https://godlike.com.au/fileadmin/godlike/techtools/ctrlr/binaries/CTRLR_v5.6.23_resizable_window.zip

The only resources about Ctrlr for the moment are here on this github discussions or on the older ctrlr.org forum. 

The 5.6.23 build available there has many bug from the 5.3.198 fixed and has the new JUCE graphical interface and color scheme.
Hope it helps. Don't hesitate to ask anything ;)
Cheers
Damien

**Reply by puffer3** [reply]
Thanks so much!  I am happy to finally be able to ask some questions after working with this program for a month or so.

I am making some panels and have them looking nice and working as they should. I am building panels for a friend and I know that they will want to be able to right click modulators in the DAW/.vst final and be able to add automation like in a regular plug in. The destined daw for these panels is Fruity Loops.

So far I have only been able to add modulation by browsing the modulator list add adding automation that way. The knobs in the panel respond to changes from the DAW so that is functioning. My question is: Is there some sort of right click enable that makes the modulators behave as other parameters in a DAW? Or is this wrapper isolated from the DAW in a way that prohibits that?

**Comment by damiensellier** [comment]
EDIT

I made a demo panel with a "pseudo-tooltip" function on mouse-over
it's working great!

the content of the tooltip is coming from the "hovered-on" modulator property ModulatorCustomName.

The hover make the tooltip visible and it vanishes after the decay time set in the timer (ex 250ms)

Hope it helps


![ctrlr-hover-tooltip](https://user-images.githubusercontent.com/55164385/156862368-fd28c291-d4f9-4002-8478-3249686ebfb1.gif)



[PseudoTooltip_1_0_0_2022-03-04_18-31.panel.zip](https://github.com/RomanKubiak/ctrlr/files/8189530/PseudoTooltip_1_0_0_2022-03-04_18-31.panel.zip)


PS : Ctrlr and any plugin works as an app within the host app so their functions are independent. But Ctrlr communicates with the host and passes the modulator values to the host like FL Studio if the property "export parameter to vst host" is enabled for each sliders.

PS2 : Everything with my pseudo-tooltip demo is working via lua scripting. If you need some help to understand how it works just let me know.

**Reply by Ted-juh** [reply]
Right-click is available as long as you can attach a mouseDown LUA script to the Component. 

eventMods = event.mods:getRawFlags()

if eventMods == 32 then
-- do something
end

32 is the right-mouse-button. 16 would be left-mouse-button. 

I was about to tell you about BubbleMessageComponent but Dnaldoog beat me to it.

**Reply by dnaldoog** [reply]
I didn't know about that one! Here is a panel that only triggers on right click.

<pre>
--
-- Called when a mouse is down on this component
--

showMyBubble = function(--[[ CtrlrComponent --]] comp --[[ MouseEvent --]], event)
    local compName = L(comp:getOwner():getName())
    local eventMods = event.mods:getRawFlags()
   if eventMods == 32 then
        --32 is the right-mouse-button. 16 would be left-mouse-button.
        -- or ModifierKeys.rightButtonModifier ModifierKeys.leftButtonModifier
        if compName == "mySlider" then
            local b = BubbleMessageComponent(500)
            b:showAt(panel:getComponent(compName), "set LFO rate", 2500, true, false)
        elseif compName == "myButton" then
            local b = BubbleMessageComponent(500)
            b:showAt(panel:getComponent(compName), "Please click here to save", 500, true, false)
        end
    end -- eventmods
end 
--[[ https://docs.juce.com/master/classBubbleMessageComponent.html --]]
</pre>

----

[ToolTips for Modulators_2_0.zip](https://github.com/RomanKubiak/ctrlr/files/8190393/ToolTips.for.Modulators_2_0.zip)

**Reply by Ted-juh** [reply]
For the rest of the codes:

16 LMB
17 LMB + Shift
18 LMB + Ctrl
19 LMB + Shift + Ctrl
20 LMB + Alt
21 LMB + Alt + Shift
22 LMB + Alt + Ctrl	
23 LMB + Alt + Shift + Ctrl

64 MMB
65MMB + Shift
66 MMB+Ctrl
67 MMB+Shift+ Ctrl
68 MMB+Alt
69 MMB+Alt+Shift
70 MMB+Alt+Ctrl
71MMB+Alt+Shift

For RMB:
32 RMB
33 RMB+Shift
34 RMB+Ctrl
35 RMB+Shift+Ctrl
36 RMB+Alt
37 RMB+Alt+Shift
38 RMB+Alt+Ctrl
39 RMB+Alt+Shift

**Comment by dnaldoog** [comment]
There is also a built in use of [JUCE **BubbleMessageComponent**](https://docs.juce.com/master/classBubbleMessageComponent.html) available in Ctrlr. 


[](url)

Here is the code in Lua:

<pre>
--
-- Called when a mouse is down on this component
--

showMyBubble = function(--[[ CtrlrComponent --]] comp --[[ MouseEvent --]], event)
    local compName = L(comp:getOwner():getName())
    if compName == "mySlider" then
        local b = BubbleMessageComponent(500)
        b:showAt(panel:getComponent(compName), "set LFO rate", 2500, true, false)
    elseif compName == "myButton" then
        local b = BubbleMessageComponent(500)
        b:showAt(panel:getComponent(compName), "Please click here to save", 500, true, false)
    end
end

--[[ https://docs.juce.com/master/classBubbleMessageComponent.html --]]
</pre>
![mouseDown440](https://user-images.githubusercontent.com/76831683/156875547-9925743e-5735-430c-a3ba-af76f90d1acd.png)

![mouseDown440(2)](https://user-images.githubusercontent.com/76831683/156875499-1fa1777c-ca9e-4e16-9a48-2ab1847fde64.png)



[ToolTips for Modulators_1_0.zip](https://github.com/RomanKubiak/ctrlr/files/8190304/ToolTips.for.Modulators_1_0.zip)

**Reply by damiensellier** [reply]
I didn't know it was possible. My pseudo-tooltip feel sorry for itself compared to the stock BubbleMessageComponent...
 I can see you both discovered hidden features recently :)
Thanks guys!

**Reply by dnaldoog** [reply]
@damiensellier  Bon Jour Damien!

That panel wasn't working on v __5.3.201__ so I made some changes: I think the mouse hover was overloading the system and crashing ctrlr, so I put a brake on it using a metatable as a flag. It was the first thing I thought of and I don't know if it's the best way, but instead of using  `allowTooltip` as a global variable I use a metatable to reset and/or block the hover from triggering the function more than once.

I didn't know about `panel:getModulatorByName("tooltip"):getComponent():setTopLeftPosition(ModPositionX - 10, ModPositionY - 30)` - very nice.

<pre>
myMethod = function(--[[ CtrlrComponent --]] comp --[[ MouseEvent --]], event)
    local ModName = L(comp:getOwner():getName())
    t[ModName] = t[ModName] + 1

    if t[ModName] == 1 then
        console("you just hovered a modulator " .. ModName)
        ParamModCustomName = panel:getModulatorByName(ModName):getProperty("modulatorCustomName")
        ModPositionX = comp:getX()
        ModPositionY = comp:getY()
        panel:getModulatorByName("tooltip"):getComponent():setTopLeftPosition(ModPositionX - 10, ModPositionY - 30)
        -- Set X,Y Position on top of the tooltip X-10px Y-30px
        if ParamModCustomName ~= nil then
            console("ParamModCustomName : " .. ParamModCustomName)
            panel:getLabel("tooltip"):setText(ParamModCustomName) 
            -- pass the modulatorCustomName as tooltip content
        end

        panel:getModulatorByName("tooltip"):getComponent():setProperty("componentVisibility", 1, false)
        -- turns the tooltip visible

        timer:setCallback(1, TimerCallback) -- TimerId 1
        timer:startTimer(1, 3500) -- Timer Id  3500ms release
    end

    for k, v in pairs(t) do
        if k ~= ModName then
            t[k] = nul
        end
    end
end
</pre>

----

<pre>
	t=setmetatable({},{__index=function(t,k)
	t[k]=0
		return t[k]
	end
	})
</pre>
 
----

<pre>
TimerCallback = function(timerId)
    if timerId == 1 then
        panel:getModulatorByName("tooltip"):getComponent():setProperty("componentVisibility", 0, false)
        timer:stopTimer(timerId)
    end
end -- f
</pre>

----

[PseudoTooltip_2_0.zip](https://github.com/RomanKubiak/ctrlr/files/8190498/PseudoTooltip_2_0.zip)

**Reply by dnaldoog** [reply]
Actually I just tried streamlining the _metatable_ especially the `t[ModName] = t[ModName] + 1` code but it was crashing - would have to give it deeper thought to make it more efficient, but this version seems to be stable!

**Reply by Ted-juh** [reply]
You are correct, it "is" overloading the system. Well, the console of Ctrlr that is. You're basically asking Ctrlr to give a console message many times a second to say: ("you just hovered a modulator " .. ModName). Ctrlr can't handle that and crashes or outputs thousands of messages after a longgggg time.

Funny fact, if you update a label with that message it should be butter smooth. At least that's what I found out the hard way.

**Reply by dnaldoog** [reply]
> You are correct, it "is" overloading the system. Well, the console of Ctrlr that is. You're basically asking Ctrlr to give a console message many times a second to say: ("you just hovered a modulator " .. ModName). Ctrlr can't handle that and crashes or outputs thousands of messages after a longgggg time.
> 
> Funny fact, if you update a label with that message it should be butter smooth. At least that's what I found out the hard way.

I disabled all the comments but it still locks up!

**Comment by dnaldoog** [comment]
I rewrote the panel  and changed the metatable to a true/false boolean.

[PseudoTooltip_4_0_0.zip](https://github.com/RomanKubiak/ctrlr/files/8192541/PseudoTooltip_4_0_0.zip)

<pre>
myMethod = function(--[[ CtrlrComponent --]] comp --[[ MouseEvent --]], event)
    local ModName = L(comp:getOwner():getName())
    t[ModName] = true
        console("you just hovered a modulator " .. ModName)
        ParamModCustomName = panel:getModulatorByName(ModName):getProperty("modulatorCustomName")
        ModPositionX = comp:getX()
        ModPositionY = comp:getY()
        panel:getModulatorByName("tooltip"):getComponent():setTopLeftPosition(ModPositionX - 10, ModPositionY - 30) 
       -- Set X,Y Position on top of the tooltip X-10px Y-30px
        if ParamModCustomName ~= nil then
            console("ParamModCustomName : " .. ParamModCustomName)
            panel:getLabel("tooltip"):setText(ParamModCustomName) -- pass the modulatorCustomName as tooltip content
        end

        panel:getModulatorByName("tooltip"):getComponent():setProperty("componentVisibility", 1, false)
       -- turns the tooltip visible

        timer:setCallback(1, TimerCallback) -- TimerId 1
        timer:startTimer(1, 3500) -- Timer Id  1200ms release

    	for k, v in pairs(t) do
        if k ~= ModName then
            t[k] = nul
      end
    end

end
</pre>

----

<pre>
	t=setmetatable({},{__index=function(t,k)
	t[k]=false
		return t[k]
	end
	})
</pre>
 
----

<pre>
TimerCallback = function(timerId)
    if timerId == 1 then
        panel:getModulatorByName("tooltip"):getComponent():setProperty("componentVisibility", 0, false)
        timer:stopTimer(timerId)
    end
end -- f
</pre>

----

**Reply by damiensellier** [reply]
Wow thanks John! I read about the LUA metatable before but never really understood when to use it and there you go!

You are right, I forgot to comment the console lines and it overloads the app. Mine didn't crash though.
The hover trigger is sent every 10ms I think, so it's a lot of console messages to display in a second, hence the overload :)

I really appreciate how you cleaned up my demo, I wrote it pretty fast when the OP asked but your spin on it is so much better.
I'll use the BubbleMessageComponent anyway when required. 
Thanks, merci!
Damien

**Reply by dnaldoog** [reply]
Hi Damien,

Yes once you get your head around `setmetatable` it's very useful. I guess it's just another way of doing things!

For example you could have a default message for all modulators, but maybe there are one or two that have different messages.
<pre>
t=setmetatable({lfo="this is my LFO",vcf="I like to use vcf sometimes"},{__index=function(t,k)
t[k]="You clicked on "..k
return t[k]
end


printToLabel = function(--[[ CtrlrModulator --]] mod, --[[ number --]] value, --[[ number --]] source)

local name=L(mod:getName())
 panel:getLabel("out"):setText(t[name])

end

</pre>

The metatable is never called for vcf and lfo.
The important thing to understand is that once the metatable is called (say you change the value for the chorus button in the example panel) it won't be called again as it sets the value in table` t` (in memory). In other words table `t` will now be:

<pre>
t={lfo="this is my LFO",vcf="I like to use vcf sometimes",chorus="You clicked on chorus"}
</pre>

and so on: You could have 50 modulators and this code wouldn't change if they all shared the default message apart from lfo and vcf.
You can see that there really isn't much code! 😍

[metatable example_1_0_0_2022-03-06_21-37.zip](https://github.com/RomanKubiak/ctrlr/files/8192592/metatable.example_1_0_0_2022-03-06_21-37.zip)


----

P.S. I changed the code in the previous post above by taking out the unnecessary `if t[ModName] then` because it is always set to true in the immediately preceding line ... duh 😒 - so uploaded a new panel in that post.

**Reply by damiensellier** [reply]
That is so nice! This way you don't have to call different scripts depending on the modulator and just have one with exceptions, that is neat. I can already see some use for that ;)
Thanks for the tip and have a nice week.
Damien

**Reply by puffer3** [reply]
How can I modify this amazing and simple code to have it function on mouseover instead of click?  Just changing the function execution to when mouse is over component gives me the error
![Screen Shot 2022-04-12 at 6 49 30 PM](https://user-images.githubusercontent.com/6006231/163083689-65347f70-0dd3-41af-8bd4-d53f677a229d.jpg)

**Reply by Ted-juh** [reply]
Attach the script to the mouseOver callback and change: 
local name = L(mod:getName())
into: 
local name=L(comp:getOwner():getName())

And it should work.

**Reply by puffer3** [reply]
what does the comp mean?  I am still getting an error that error but with 'comp' instead of "getName'

**Reply by Ted-juh** [reply]
Are you able to zip the panel and post it? It worked just fine when I changed Dnaldoog's panel.

To make sure: You created a new script on the mouseOver callback for each modulator (not the valueChanged method)? 

Mod stands for modulator. Comp stands for component.

**Reply by puffer3** [reply]
Here is my panel @Ted-juh 

Thanks! is there somewhere I can refer to that has some syntax that would help with learning these kinds of things?

[metatable_Mouseover.bpanelz.zip](https://github.com/RomanKubiak/ctrlr/files/8484819/metatable_Mouseover.bpanelz.zip)

**Reply by dnaldoog** [reply]
Hi Henry,

When you create a ' mouse' event function, you'll notice the function parameters are `(comp`,`event)`.
When you create a value change function, you'll see `(mod,value,source)`.

The object is `mod` or `comp` depending. You can access the `mod` from `comp` using `getOwner()` - it's confusing!

So , if you want a message on mouseOver you need a **new function**

<pre>
--
-- Called when the mouse moves over a component
--
printToLabelHover = function(--[[ CtrlrComponent --]] comp, --[[ MouseEvent --]] event)
printToLabel(comp:getOwner(),nil,nil)
end
</pre>

[metatable example_2_0_with_hover.zip](https://github.com/RomanKubiak/ctrlr/files/8485009/metatable.example_2_0_with_hover.zip)

**Reply by Ted-juh** [reply]
Just a question out of curiousity. The mouseOver means it generates a lot of calls and would mean the label (or tooltip) gets updated that amount of times.

To my knowledge it's better to just use mouseEnter and mouseExit because it would generate just two calls. But I couldn't find a way to do it because of missing callbacks.

Is there a way to do it or is the mouseOver function easy to handle (gui-wise) for Ctrlr?

**Reply by damiensellier** [reply]
@Ted-juh there's additional MouseListener callbacks available such as mouseEnter() and mouseExit() associated to uiCustomComponent.
![Capture d’écran, le 2022-04-14 à 15 03 06](https://user-images.githubusercontent.com/55164385/163458950-7662aaa1-579b-44eb-94d4-eec53a7e059d.png)

I tried the mouseExit() it's working but triggers several times when you leave the component's space.

**Reply by Ted-juh** [reply]
I know it's available in the Custom Component. But both mouseEnter and MouseExit aren't available for a slider itself.  While they are in the juce slider class. Could be that they became available later on. And yes, you are right about the fact that MouseEnter triggers a few times but not as much as mouseMove. (Weird, because one would assume it only triggers once.)

**Comment by puffer3** [comment]
Thanks for all these tips! I will try them all out.  

Sorry to combine requests.  My desire to be able to right click in the plug is just so the user can assign automation to specific modulators in Fruity Loops.  This looks like it will be a pain.  I know there is a modulators list where you can just browse and assign but I know my user will not want to take that extra step.

I am going to try and get some tool tip on mouse over going using these hints...or tips for tool tips tips.  Will report back. 

I will post my next question as a new thread to help avoid confusion.

You all rock

**Reply by damiensellier** [reply]
I don't understand what you try to achieve.
Is your Ctrlr plugin like a virtual bank of modulators to control other parameters in FL?
Or you just want to record in FL the slider moves from Ctrlr to send midi to an external device?

**Reply by puffer3** [reply]
Sorry it is a bit confusing i know.  I am making a panel for a vintage synth and in fruity loops when you want to add modulation to a parameter you right click and theres a menu to "add automation clip" for the selected parameter.  

after further research (I am fruity loops noob) it appears no 3rd party plug lets u access this menu.  The best bet is to create a keyboard shortcut for "options>last tweaked>add automation clip" It seems.  Using autohotkey, another program idk haha.

Tools tips stuff is great! Thanks everyone for posting.

I am going to have to do some tricky sysex stuff coming up so I am sure I will be here asking more questions.

**Comment by puffer3** [comment]
Ok I am working on this project some.  I am changing the way I approach the tool tips to the style that valhalla uses in their reverb plugins.  So what I am after is a single label at a single location that gets written to when the mouse is over certain components.  I am going to try and use the metatable style and send text to the label but I am sure I will need help lol so stay tuned.  You all have been a great and patient community so far.

**Comment by dnaldoog** [comment, marked as answer]
This panel shows how to use _BubbleMessageComponent_:

**edit**: Looks like I've already explained this ― completely forgot. Oh well.

![BubbleMessageComponentExample](https://user-images.githubusercontent.com/76831683/163924613-bf59b25f-4bfb-4892-b256-ba59fdf954ba.png)

[metatable example bubble.zip](https://github.com/RomanKubiak/ctrlr/files/8510003/metatable.example.bubble.zip)

**Comment by puffer3** [comment]
Ok I have my tool tips window working right but it doesn't seem to update as fast as the mouse moves.  Maybe the mouseEnter and mouseExit calls would be better? or is this just my mac version of ctrlr being wack?
[OB8patchNameTest.bpanelz.zip](https://github.com/RomanKubiak/ctrlr/files/8525266/OB8patchNameTest.bpanelz.zip)
?

**Comment by dnaldoog** [comment]
Your panel is generating a huge amount of mousehovers over a control. If the user were changing values constantly, there would be hundreds of messages generated. This code fixes that.

<pre>
printToLabelHover = function(--[[ CtrlrComponent --]] comp --[[ MouseEvent --]], event)
    local n = L(comp:getOwner():getName())
    if dontrepeat == n then
        return
    end
    printToLabel(comp:getOwner(), nil, nil)
    dontrepeat = n
end
</pre>

Also, on Windows the label font size 12 is too big, so I dropped it to 11!
</pre>

[OB8_3_0_OB-8DAGDD_hover brake.zip](https://github.com/RomanKubiak/ctrlr/files/8525946/OB8_3_0_OB-8DAGDD_hover.brake.zip)

**Reply by puffer3** [reply]
Thanks for that! It still seems like mouse hover gets stuck on certain values and doesnt refresh.  the LFO and LFO waveshape buttons dont trigger the tool tips window at all for some reason.  Is there something i am missing?

**Reply by dnaldoog** [reply]
What version of Ctrlr are you on?

**Comment by dnaldoog** [comment]
1. `uiImageButton` is buggy and doesn't work with mouse functions; they only work when the mouse is over the label - In your LFO waveforms, you have hidden the label.
2. The large LFO button `lfoFreq` doesn't have the  _mouse moved over_ `printToLabelHover` callback function set.
3. You have a spelling mistake with: `modTri = "Selects LFO Triange Waveform",`

What you can do, as a hack, is have invisible uiButtons in the layer above named __modTri__ etc and then trigger the uibutton with a name like __modTriDummy__  sitting in the layer below to light up, using code like:

<pre>
--
-- Called when a mouse is down on this component
--

myMouseDown = function(--[[ CtrlrComponent --]] comp, --[[ MouseEvent --]] event)

local n=L(comp:getOwner():getName())
panel:getLabelComponent("d"):setText("you clicked on \nmodulator \187 "..n)

if n=="modTri" then
local val=comp:getOwner():getValue()
panel:getModulatorByName("modTriDummy"):getComponent():setValue(val,true)
end
end
</pre>

![uibuttonmouseoverfail](https://user-images.githubusercontent.com/76831683/164592226-c646513f-22d5-49a2-8d9e-971efd9b3aca.png)

----

**EDIT 2022/05/13 added bpanelz file (previous panel file had no resources included)**

[mouseOverImageButton_1_0.zip](https://github.com/RomanKubiak/ctrlr/files/8683219/mouseOverImageButton_1_0.zip)
